kimchi/src/folding/expressions.rs

```diff
@@ -0,0 +1,406 @@
+            FoldingExp::Add(e1, e2) => {
+                let e1 = Box::new(e1.into_compatible());
+                let e2 = Box::new(e2.into_compatible());
+                BinOp(Op2::Add, e1, e2)
+            }
+            FoldingExp::Sub(e1, e2) => {
+                let e1 = Box::new(e1.into_compatible());
+                let e2 = Box::new(e2.into_compatible());
+                BinOp(Op2::Sub, e1, e2)
+            }
+            FoldingExp::Mul(e1, e2) => {
+                let e1 = Box::new(e1.into_compatible());
+                let e2 = Box::new(e2.into_compatible());
+                BinOp(Op2::Mul, e1, e2)
+            }
+        }
+    }
+}
+impl std::ops::Add for Degree {
+    type Output = Self;
+
+    fn add(self, rhs: Self) -> Self::Output {
+        use Degree::*;
+        match (self, rhs) {
+            (Zero, Zero) => Zero,
+            (Zero, One) | (One, Zero) | (One, One) => One,
+            (_, Two) | (Two, _) => Two,
+        }
+    }
+}
+impl std::ops::Mul for &Degree {
+    type Output = Degree;
+
+    fn mul(self, rhs: Self) -> Self::Output {
+        use Degree::*;
+        match (self, rhs) {
+            (Zero, Zero) => Zero,
+            (Zero, One) | (One, Zero) => One,
+            (Zero, Two) | (One, One) | (Two, Zero) => Two,
+            (One, Two) | (Two, One) | (Two, Two) => panic!("degree over 2"),
+        }
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct Term<C: FoldingConfig> {
+    pub exp: FoldingExp<C>,
+    pub sign: bool,
+}
+
+impl<C: FoldingConfig> Term<C> {
+    /*fn map_exp<F>(self, f: F) -> Self
+    where
+        F: Fn(FoldingExp2<C>) -> FoldingExp2<C>,
+    {
+        let Self { exp, sign } = self;
+        let exp = f(exp);
+        Self { exp, sign }
+    }*/
+    fn wrap_exp<F>(self, f: F) -> Self
+    where
+        F: Fn(Box<FoldingExp<C>>) -> FoldingExp<C>,
+    {
+        let Self { exp, sign } = self;
+        let exp = f(Box::new(exp));
+        Self { exp, sign }
+    }
+}
+impl<C: FoldingConfig> std::ops::Mul for &Term<C> {
+    type Output = Term<C>;
+
+    fn mul(self, rhs: Self) -> Self::Output {
+        #[allow(clippy::suspicious_arithmetic_impl)]
+        let sign = !(self.sign ^ rhs.sign);
+        let exp = FoldingExp::Mul(Box::new(self.exp.clone()), Box::new(rhs.exp.clone()));
+        Term { exp, sign }
+    }
+}
+
+impl<C: FoldingConfig> std::ops::Neg for Term<C> {
+    type Output = Self;
+
+    fn neg(self) -> Self::Output {
+        Term {
+            sign: !self.sign,
+            ..self
+        }
+    }
+}
+///A simplified expression with all terms separated by degree
+#[derive(Clone, Debug)]
+pub struct IntegratedFoldingExpr<C: FoldingConfig> {
+    //(exp,sign,alpha)
+    pub(super) degree_0: Vec<(FoldingExp<C>, bool, usize)>,
+    pub(super) degree_1: Vec<(FoldingExp<C>, bool, usize)>,
+    pub(super) degree_2: Vec<(FoldingExp<C>, bool, usize)>,
+}
+
+impl<C: FoldingConfig> Default for IntegratedFoldingExpr<C> {
```

Comment:
It seems likely that you could `derive(Default)` for this (but apologies if not).